### PR TITLE
Add newRedfish role OemIBMServiceAgent

### PR DESCRIPTION
**Pull request by joseph-reynolds**
This adds a new OEM Redfish privilege OemIBMPerformService.  This is needed
to mark REST API operations which should be restricted to the service user.

This adds a new custom Redfish role OemIBMServiceAgent which corresponds to
the priv-oemibmserviceagent role in phosphor-user-manager.  This role has
all privileges including the new OemIBMPerformService privilege.

Tested:
Tested only local users, not LDAP RemoteRoleMapping.
Tested on a system which had the new OemIBMServiceAgent role.

Setup: Ensure there is a "service" use who has role=OemIBMPerformService.
How?  Log into the BMC command shell as the root user:
  useradd service -G priv-oemibmserviceagent,web,redfish \
    -m -N -s /bin/sh -e ""
  passwd service   ---enter the password twice---
  systemctl restart xyz.openbmc_project.User.Manager.service

Tested aspects of DSP0266 version 1.12.0 "Restricted roles and restricted
privileges".  Numbers here correspond to bullets in that section:

1a. POST /redfish/v1/AccountService/Accounts/new with Role:OemIBMServiceAgent
Ensure it failed with message Base.1.9.0.RestrictedRole.

1b. PATCH /redfish/v1/AccountService/Accounts/ordinary with
Role:OemIBMServiceAgent.  Ensure it failed.

2. PATCH the "service" user (all property) and ensure it fails for all
properties except "Locked".

3. DELETE the "service" user && ensure it failed

4. PATCH the OemIBMServiceAgent role as the LocalRole within the
RemoteRoleMapping property && ensure it failed.

IMPI: Use ipmitool to try to create or modify the service user:
ipmitool user set name ...
ipmitool user set password ...

Signed-off-by: Joseph Reynolds <joseph-reynolds@charter.net>

Required prerequisite: the openbmc and phosophor-user-manager commits which create the corresponding new oemibmserviceagent Phosphor Privilege Role.

Note: This add the OemIBMServiceAgent Role, OemIBMPerformService Privilege, and enforces that OemIBMServiceAgent is a Restricted Role.  It does not require the OemIBMPerformService privilege for any operation.

### CODE REVIEW

File: redfish-core/lib/roles.hpp [hunk 1]
```diff
@@ -64,13 +68,57 @@ inline bool getAssignedPrivFromRole(std::string_view role,
     {
         privArray = nlohmann::json::array();
     }
+    else if (role == "OemIBMServiceAgent")
+    {
+        privArray = {"Login", "ConfigureManager", "ConfigureUsers",
+                     "ConfigureSelf", "ConfigureComponents"};
```
Comment (human3rr) [inline redfish-core/lib/roles.hpp:74]: Is there not an additional priv to add here (OemIBMPerformService)?
Please clarify if I'm misunderstanding.

Output from get role oemIBMServiceAgent:
curl -k -H "X-Auth-Token: $token" -X GET https://${bmc}/redfish/v1/AccountService/Roles/OemIBMServiceAget
{
  "@odata.id": "/redfish/v1/AccountService/Roles/OemIBMServiceAgent",
  "@odata.type": "#Role.v1_2_2.Role",
  _"AssignedPrivileges": [
    "Login",
    "ConfigureManager",
    "ConfigureUsers",
    "ConfigureSelf",
    "ConfigureComponents"
  ],_
  "Description": "OemIBMServiceAgent User Role",
  "Id": "OemIBMServiceAgent",
  "IsPredefined": true,
  "Name": "User Role",
  "OemPrivileges": [
    "OemIBMPerformService"
  ],
  "Restricted": true,
  "RoleId": "OemIBMServiceAgent"

**Review by human3rr — COMMENTED** [review]
"Ensure there is a "service" **use**"
assume typeo for use, should be user

I've looked over all the changes and other than the other comment I made I don't see any issues.  Let me know about my comment and I'll approve this and corequisites

File: redfish-core/lib/roles.hpp (hunk 1 above)
Comment (joseph-reynolds) [reply redfish-core/lib/roles.hpp:74]: Yes, that is intended.  Redfish separates Redfish standard privileges and OEM privileges.

**Review by human3rr — APPROVED** [review]
Looks good

**Comment by joseph-reynolds** [comment]
> Ensure there is a "service" **use**

Yes, it is a typo.   Should be: "service" user.

**Review by geissonator — CHANGES_REQUESTED** [review]
This is failing CI, looks like clang formatting, please take a look.

File: redfish-core/include/privileges.hpp [hunk 2]
```diff
@@ -224,6 +224,13 @@ inline const Privileges& getUserPrivileges(const std::string& userRole)
         static Privileges readOnly{"Login", "ConfigureSelf"};
         return readOnly;
     }
+    if (userRole == "priv-oemibmserviceagent")
+    {
+        static Privileges admin{"Login", "ConfigureManager", "ConfigureSelf",
+                                "ConfigureUsers", "ConfigureComponents",
+                                "OemIBMPerformService"};
+        return admin;
```
Comment (gtmills) [inline redfish-core/include/privileges.hpp:232]: This doesn't look correct

File: redfish-core/include/privileges.hpp [hunk 3]
```diff
@@ -224,6 +224,13 @@ inline const Privileges& getUserPrivileges(const std::string& userRole)
         static Privileges readOnly{"Login", "ConfigureSelf"};
         return readOnly;
     }
+    if (userRole == "priv-oemibmserviceagent")
+    {
+        static Privileges admin{"Login", "ConfigureManager", "ConfigureSelf",
```
Comment (gtmills) [inline redfish-core/include/privileges.hpp:229]: static Privileges ibmserviceagent ?

**Review by gtmills — CHANGES_REQUESTED** [review]
Can you run the Redfish validator ?

Need clang-formated

File: redfish-core/include/privileges.hpp [hunk 4]
```diff
@@ -155,7 +155,7 @@ class Privileges
         size_t endIndex = basePrivilegeCount;
         if (type == PrivilegeType::OEM)
         {
-            searchIndex = basePrivilegeCount - 1;
+            searchIndex = basePrivilegeCount;
```
Comment (gtmills) [inline redfish-core/include/privileges.hpp:158]: Why change this?

**Review by gtmills — COMMENTED** [review]
Should you do some unit tests?
https://github.com/openbmc/bmcweb/blob/9a128eb3fc3008f802974360e98de6495ed1a66b/redfish-core/ut/privileges_test.cpp

File: redfish-core/include/privileges.hpp (hunk 4 above)
Comment (joseph-reynolds) [reply redfish-core/include/privileges.hpp:158]: Oops.  This is to fix a code bug in handling OEM privileges (which BMCWeb does not yet have upstream).  Note that basePrivilegeCount is 5, and the search for OEM privileges should start at index 5, not 4.  Off by one.
I created https://gerrit.openbmc-project.xyz/c/openbmc/bmcweb/+/46154 to merge this upstream.

File: redfish-core/include/privileges.hpp (hunk 3 above)
Comment (joseph-reynolds) [reply redfish-core/include/privileges.hpp:229]: Oops.  The variable name is wrong.  The error has no effect on the code.  I'll change it in the next commit.

File: redfish-core/include/privileges.hpp (hunk 2 above)
Comment (joseph-reynolds) [reply redfish-core/include/privileges.hpp:232]: Can you be more specific?  I'll correct the variable name per the previous comment.  And the set of privileges is the same as admin, plus the OemIBMPerformService privilege.

**Comment by joseph-reynolds** [comment]
> Should you do some unit tests?

I performed testing outside the gmock framework.  I'll try to run the tests and enhance if needed.

File: redfish-core/include/privileges.hpp (hunk 2 above)
Comment (gtmills) [reply redfish-core/include/privileges.hpp:232]: yeah sorry, i should have been more specific. 
Can you follow the pattern and call this something like ibmserviceagent

File: redfish-core/include/privileges.hpp (hunk 4 above)
Comment (gtmills) [reply redfish-core/include/privileges.hpp:158]: Cool. Thanks for doing that. I merged https://gerrit.openbmc-project.xyz/c/openbmc/bmcweb/+/46154

**Comment by rfrandse** [comment]
Outdated - PR older than two months

**Pull request closed without merging**